### PR TITLE
Add Model-Owned Extensions to AFD P2P Connectors

### PR DESCRIPTION
**Pull request by lirx-pd**
## Purpose

Add a model-owned extension mechanism to AFD connectors so new model architectures can transfer additional Attention-to-FFN control and data-plane information without adding model-specific fields and transport logic to every connector.

This change introduces a no-op `AFDConnectorExtension` in the common connector layer. Models may provide an implementation through `get_afd_connector_extension()`, while models without an extension preserve the existing communication behavior.

The proposal is described in **[RFC] Add Model-Owned Extensions to AFD Connectors**.

## Issue

- Related issue(s): #209 
- Closing keyword, only if fully resolved: N/A #209 

## Scope

- In scope:
  - Add the plugin-owned `AFDConnectorExtension` Null Object and model binding entry point.
  - Add an optional opaque extension object to the AFD control payload.
  - Add an optional opaque extension result to Attention-to-FFN transfer metadata.
  - Invoke extension hooks from GPU P2P, synchronous NPU CAMP2P, and asynchronous NPU CAM connectors.
  - Bind model-owned extensions after model loading in GPU and NPU Attention/FFN runners.
  - Preserve the current behavior for models that do not provide an extension.
  - Add unit coverage for binding, control-payload serialization, control-plane delegation, and data-plane hooks.
- Out of scope:
  - Model-specific extension implementations such as DeepSeek V4 `input_ids` transport.
  - A generic tensor schema, dtype/shape registry, or connector-managed extension buffers.
  - Per-layer extension schemas; one model-owned extension is used for the complete model.
  - Changes to vLLM source code or public vLLM interfaces.
  - Real multi-process GPU or NPU communication validation in this PR.

## Implementation Notes

- `AFDConnectorExtension` is defined in the plugin connector base module and exported from `afd_plugin.connectors`.
- `AFDConnectorBase` creates a no-op extension by default. `bind_model()` replaces it only when the loaded model exposes `get_afd_connector_extension()`.
- The extension interface contains four hooks:
  - `control_payload()` produces optional model-owned JSON control data.
  - `update_state_from_control_payload()` applies received control data.
  - `send_attn_extention()` sends model-owned Attention-to-FFN data after the connector's standard payload.
  - `recv_attn_extention()` receives model-owned data and returns the value exposed to FFN model code.
- `AFDControlPayload.extension` transports opaque JSON-compatible control data through the existing plugin-owned serializer.
- `AFDTransferMetadata.extension` carries the opaque data-plane result through the existing FFN forward context, avoiding runner-specific model arguments.
- GPU P2P and synchronous NPU CAMP2P invoke both control-plane and data-plane hooks.
- Asynchronous NPU CAM has no DP metadata control plane and therefore invokes only data-plane hooks.
- No model-specific tensor name, dtype, shape, validation, or buffer policy is added to the common connector layer.
- The implementation remains entirely plugin-owned and does not add compatibility patches or monkey patches.

## Test Plan

Static and CPU-safe unit validation:
Accelerator-gated validation:

- Run GPU P2P Attention/FFN multi-process smoke coverage when a CUDA/NCCL environment is available.
- Run CAMP2P and asynchronous CAM smoke coverage when a torch-npu/CANN environment is available.
- Add model-specific end-to-end coverage together with the first model extension implementation.

## Test Result

- Ruff lint: passed.
- Ruff format check: passed.
- Connector, GPU/NPU runner, and package unit suite: passed; optional accelerator-dependent tests were skipped when their runtime was unavailable.
- Python compilation checks: passed.
- `git diff --check`: passed.
- Real multi-process GPU P2P E2E: not run; no CUDA/NCCL deployment was available in the validation environment.
- Real NPU CAMP2P/Async CAM E2E: not run; no torch-npu/CANN deployment was available in the validation environment.

## Docs Impact

- Files updated: None.
- If none, reason: The public extension contract is documented by English class and method docstrings in the plugin-owned connector modules. The linked RFC provides the design rationale. Model-specific usage documentation should be added with the first model extension implementation.

---

<details>
<summary>Essential PR Checklist</summary>

- [x] Purpose is clear and linked to public context when possible.
- [x] Scope is bounded.
- [x] Compatibility with vLLM v0.26.0 is considered.
- [x] No changes are made to the vLLM source checkout.
- [x] Plugin-owned classes or explicit dotted class paths are preferred over monkey patches.
- [x] Any compat shim or monkey patch is isolated, idempotent, version-guarded, documented, and tested.
- [x] Imports remain CPU-safe; CUDA-heavy work is delayed or GPU-gated.
- [x] Validation evidence is included, including skipped GPU tests when applicable.
- [x] Documentation impact is stated.

</details>

### CODE REVIEW

File: afd_plugin/connectors/base.py [hunk 1]
```diff
@@ -34,6 +34,39 @@ def to_mapping(self) -> dict[str, Any]:
...
+        return None
+
+    def update_state_from_control_payload(
+        self,
+        connector: AFDConnectorBase,
+        payload: dict[str, Any] | None,
+    ) -> None:
+        pass
+
+    def send_attn_output(
```
Comment (specture724) [inline afd_plugin/connectors/base.py]: send_attn_extention

File: afd_plugin/connectors/base.py [hunk 2]
```diff
@@ -34,6 +34,39 @@ def to_mapping(self) -> dict[str, Any]:
...
+        self,
+        connector: AFDConnectorBase,
+        context: AFDTransferContext,
+        **kwargs: Any,
+    ) -> None:
+        pass
+
+    def recv_attn_output(
```
Comment (specture724) [inline afd_plugin/connectors/base.py]: recv_attn_extention

**Comment by specture724** [comment]
cc @jiangkuaixue123

File: afd_plugin/connectors/base.py (hunk 1 above)
Comment (lirx-pd) [reply afd_plugin/connectors/base.py]: Already changed

File: afd_plugin/connectors/base.py (hunk 2 above)
Comment (lirx-pd) [reply afd_plugin/connectors/base.py]: Already changed

File: afd_plugin/connectors/base.py [hunk 3]
```diff
@@ -34,6 +34,39 @@ def to_mapping(self) -> dict[str, Any]:
         return {}
 
 
+class AFDConnectorExtension:
```
Comment (specture724) [inline afd_plugin/connectors/base.py:37]: Will `update_state_from_control_payload`, `send_attn_extention` and `recv_attn_extention` be the same logic to different models? If so, just make the extension dict into the payload?

File: afd_plugin/connectors/base.py [hunk 4]
```diff
@@ -34,6 +34,39 @@ def to_mapping(self) -> dict[str, Any]:
...
+        return None
+
+    def update_state_from_control_payload(
+        self,
+        connector: AFDConnectorBase,
+        payload: dict[str, Any] | None,
+    ) -> None:
+        pass
+
+    def send_attn_extention(
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/base.py:54]: **Typo: `extention` → `extension`**

`send_attn_extention` (and `recv_attn_extention` below) misspell "extension". These are public API names on the extension contract, so the typo will be locked in once a model implementation ships against them. Renaming now is the cheapest point — this is the first and only release of the interface.

Beyond the spelling, the two hooks are also asymmetric:
- `send_attn_extention(connector, context, **kwargs)` gets the full forward context and call kwargs, so a model can read `input_ids` from `kwargs` and send it.
- `recv_attn_extention(connector, ubatch_idx)` gets only `ubatch_idx` — no `context`, no way to receive anything the send side transmitted.

For a DeepSeek-V4-style `input_ids` extension this means the receive hook has no channel to read the data back from. Consider unifying the signatures (e.g. give `recv_attn_extention` the same `(connector, context, **kwargs)` shape) and defining the actual wire transport for the data-plane payload.

File: afd_plugin/connectors/base.py [hunk 5]
```diff
@@ -34,6 +34,39 @@ def to_mapping(self) -> dict[str, Any]:
...
+        self,
+        connector: AFDConnectorBase,
+        context: AFDTransferContext,
+        **kwargs: Any,
+    ) -> None:
+        pass
+
+    def recv_attn_extention(
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/base.py:62]: **Typo + `object` return type**

Same `extention` typo as `send_attn_extention`. Two additional points:

1. **Return type `object | None`** violates the project convention to avoid `object`/`Any` in public signatures (see AGENTS.md). The result is consumed as `metadata.extension = self.extension.recv_attn_extention(...)`, so a concrete protocol or at least a named type alias would be preferable.
2. **No transport channel for the data-plane payload.** `AFDTransferMetadata.extension` only carries the *receive-side* result across processes within the same worker (hidden-states transfer), but nothing moves the *send-side* data (`input_ids` from the Attention forward kwargs) to the FFN process. The send hook writes into `**kwargs` which never crosses a process boundary. If the intended DeepSeek-V4 use case is to transport `input_ids` Attention→FFN, this hook needs a real wire path (e.g. piggyback on the existing P2P tensor send, like `router_logits`), not a process-local return value.

File: afd_plugin/connectors/npu/camp2p.py [hunk 6]
```diff
@@ -638,12 +640,17 @@ def update_state_from_dp_metadata(
...
+            connector,
+            payload.extension,
+        )
 
     def send_dp_metadata_list(
         self,
         payload: AFDControlPayload,
     ) -> None:
         connector = self.connector
+        payload.extension = connector.extension.control_payload()
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/npu/camp2p.py:653]: **Control-plane callback fires on the wrong process (asymmetric)**

`update_state_from_dp_metadata` is invoked on **both** sides:
- **Attention side** (`npu/attention_model_runner.py:1384` calls `update_state_from_dp_metadata(payload)` then `send_dp_metadata_list(payload)`): applies the *Attention* process's own bound extension.
- **FFN side** (`npu/ffn_worker.py:142` `recv_dp_metadata_list()` → `execute_ffn_step` → `update_state_from_dp_metadata`): applies the *FFN* process's own bound extension.

So the `payload.extension` written by the send side is consumed by a **different process's** extension instance. If the FFN-side model does not expose `get_afd_connector_extension()` (a DeepSeek-V4-style extension may only be meaningful on the Attention side, which needs to send `input_ids`), the FFN side binds a no-op extension and `payload.extension` crosses the wire unread — while the Attention side already applied it *locally* before sending, so the send side never actually transmits its own payload.

Suggested fix: make the semantics symmetric — `send_dp_metadata_list` only **writes** `payload.extension`; the apply path (`update_state_from_control_payload`) should run only on the connector that actually **receives** (i.e. after `recv_dp_metadata_list()` decodes on the FFN side). The integration test `test_model_extension_owns_control_and_data_plane_payloads` shares one extension instance across both connectors, which masks this cross-process gap; a test with two independently-bound extensions would expose it.

File: afd_plugin/connectors/gpu/p2p.py [hunk 7]
```diff
@@ -767,6 +773,7 @@ def send_dp_metadata_list(
                 connector loop.
         """
         connector = self.connector
+        payload.extension = connector.extension.control_payload()
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/gpu/p2p.py:776]: **`payload.extension` overwrite semantics**

`send_dp_metadata_list` unconditionally overwrites `payload.extension` with `connector.extension.control_payload()`. Two consequences:

1. **Silent clobber**: if the caller constructed the payload with a pre-set `extension`, it is overwritten with no warning. `_send_dp_metadata` (`attention_model_runner.py:144`) always builds the payload fresh, so today this is benign — but the write-then-read pattern here invites the field being set from two places.
2. **No-op still writes `None`**: when the model has no extension, `control_payload()` returns `None`, and this line writes `payload.extension = None`. Combined with the symmetric apply suggestion, it would be cleaner to leave the field untouched when there is no extension (skip assignment), so `payload.extension` remains `None` from construction rather than an explicit overwrite.

Recommendation: move the extension filling into the wire-serialization boundary (`encode_control_payload`) or guard the assignment, so the control-plane extension has exactly one writer with clear ownership.

File: afd_plugin/connectors/npu/camp2p.py [hunk 8]
```diff
@@ -638,12 +640,17 @@ def update_state_from_dp_metadata(
         connector.dp_metadata_list = payload.dp_metadata_list
         connector.is_graph_capturing = payload.is_graph_capturing
         connector.is_warmup = payload.is_warmup
+        connector.extension.update_state_from_control_payload(
+            connector,
+            payload.extension,
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/npu/camp2p.py:645]: **Duplicate of the P2P control-plane asymmetry (camp2p variant)**

Same issue as my comment on `gpu/p2p.py:756`. In the camp2p path, `recv_dp_metadata_list()` (`camp2p.py:661`) decodes the wire payload but applies the extension callback only later via `update_state_from_dp_metadata`, which runs on whichever process calls it. If the symmetric semantics from the other comment are adopted, this apply should be moved to the recv path (decode → apply the *local* FFN-side extension), and the Attention-side `update_state_from_dp_metadata` (called from `npu/attention_model_runner.py:1384`) should not apply the extension it is about to send.

Keeping the two connector implementations consistent on the same control-plane contract matters here — P2P and camp2p are both control-plane-driven connectors, and diverging apply semantics between them would be a maintenance trap.
